feat: Implement infinite scroll for video list with throttle optimization

SangJin983 opened the pull request:
### 변경 사항(최초)
- YouTube API의 nextPageToken을 이용한 무한 스크롤 구현
- 스크롤 성능 최적화를 위한 throttle 함수 구현 및 적용
- 비디오 추가 로딩 기능 구현

### 세부 기술 사항
1. 핵심 기능 추가를 위한 작업
```
### 상태 관리 추가 및 수정
- `itmes -> results` 변경: fetch한 자료의 결과물을 저장한다는 의도를 보다 분명하게 나타내기 위해 수정
- `searchQuery` 추가: 비디오 추가 로딩을 위해, 사용자 검색의 결과물인 현재 query를 저장
- `appendVideos` 리듀서: 비디오를 추가하는 리듀서
```
2. 추가된 핵심 기능 설명
- 스크롤 위치 확인
```VideoList.jsx
    const scrollPosition = window.innerHeight + window.scrollY;
    const triggerPosition = document.body.offsetHeight - 500;

    // 현재 스크롤 위치가 트리거 위치에 도달했는지 확인
    if (scrollPosition >= triggerPosition) {
      loadMoreVideos();
    }
```
- 쓰로틀 함수에서의 시간 관리
```utils.js
    const now = Date.now();
    const remainingTime = wait - (now - lastCallTime); // 호출을 위해 남은 시간 확인
```
```utils.js
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn(...args);
      lastCallTime = Date.now();
    }, remainingTime); // 남은 시간을 반영한 새로운 함수 호출
```
- 비디오 목록 추가시 중복 비디오 필터링
```videoSlice.js
      const existingIds = new Set(state.results.map((v) => v.id.videoId)); // 현재 목록의 videoId 확인
      const uniqueItems = action.payload.items.filter(
        (item) => !existingIds.has(item.id.videoId)
      ); // Set의 has 기능을 통해 보다 효율적으로 중복을 확인: O(1)
```

### 추후 개선 사안 (연구 중)
- Redux thunk를 이용한 보다 중앙화된 비동기 작업 처리
- 에러 및 로딩 상태에 대한 관리 적용
- 성능 최적화를 위한 리팩토링

### 다음 브렌치 작업 예정
- thunk를 이용한 리팩토링
- Modal 창 구현 및 세부 요구사안 반영

### 참고 자료
- [Lodash의 throttle](https://lodash.com/docs/4.17.15#throttle)

SangJin983 committed:
feat: video목록 추가 기능 구현

SangJin983 committed:
Merge branch 'feat/debounce-video-search' into feat/infinite-scroll

SangJin983 committed:
feat: video목록 추가 기능 구현 및 네이밍수정
- merge된 코드에 맞춰 목록 추가를 다시 구현했다.
- 전반적인 코드의 네이밍을 좀더 분명하게 수정했다.

SangJin983 committed:
feat: 무한 스크롤 기능 구현
- throttle 함수 구현
- appendVideos 호출 시 기존의 목록과 중복된 비디오는 필터링하도록 개선
- 비디오 추가기능이 있는 handleScroll 이벤트 핸들러를 스크롤 이벤트에 등록
- fetch한 결과를 저장하는 컬렉션명을 list에서 results로 수정

pjb6510 commented on src/features/videoSlice.js:
```diff
@@ -1,20 +1,28 @@
+  name: "videos",
   initialState: {
-    items: [],
+    results: [],
```
result라고 하니 어떤 의미인지 와닿지 않네요. searchResult 정도가 어떨지요?

SangJin983 replied:
> result라고 하니 어떤 의미인지 와닿지 않네요. searchResult 정도가 어떨지요?

수정하였습니다 감사합니다.

pjb6510 commented on src/components/VideoSearch.jsx:
```diff
@@ -1,31 +1,49 @@
+  const [searchTerm, setSearchTerm] = useState("");
 
-  const handleSubmit = (e) => {
+  const excuteSearch = async (searchQuery) => {
```
`"execute"`
-
네이밍은 일관성 있게 하는게 좋을 것 같습니다. 
VideoList에서의 loadMoreVideos 동작이나 여기서의 동작에 최초 로드 여부 외에 큰 차이가 없어보입니다.
따라서 loadVideos 등으로 처리하는게 적절해보입니다.

pjb6510 replied:
이전 PR에서도 언급했듯 Query라는 단어는 어울리지 않습니다.

SangJin983 replied:
이전 PR의 내용에 대해서는, '데이터베이스에 날리는 질의문'이라는 개념을 잘 살려서 그러한 기능을 하는 함수의 매개변수에 searchQuery와 같은 네이밍을 써야 한다고 이해했는데, (그러한 까닭에 search를 실행하는 함수라는 의미의 excuteSearch로 수정해봤는데) 제가 잘못이해한 듯 합니다.
 url에 직접 입력하는 경우나 혹은 그러한 작업을 수행한 결과물을 저장하는 의미에서만 searchQuery를 사용하도록 하겠습니다.

SangJin983 replied:
> 이전 PR에서도 언급했듯 Query라는 단어는 어울리지 않습니다.

이전 PR의 내용에 대해서는, '데이터베이스에 날리는 질의문'이라는 개념을 잘 살려서 그러한 기능을 하는 함수의 매개변수에 searchQuery와 같은 네이밍을 써야 한다고 이해했는데, (그러한 까닭에 search를 실행하는 함수라는 의미의 excuteSearch로 수정해봤는데) 제가 잘못이해한 듯 합니다.

'직접적으로 fetch작업을 위해 url에 입력하는 경우'나 혹은 '그러한 작업을 수행한 결과물'을 가리키는 의미에서만 searchQuery를 사용하도록 하겠습니다.

나머지 부분도 반영하여 수정하겠습니다.

pjb6510 commented on src/components/VideoSearch.jsx:
```diff
@@ -1,35 +1,42 @@
-      fetchAndSetVideos(randomSearchTerm);
+    const excuteInitialSearch = async () => {
+      const randomTopic = getRandomYouTubeSearchTopic();
+      excuteSearch(randomTopic);
     };
 
-    fetchRandomSearchTerm();
-  }, [dispatch, generateRandomSearchTerm]);
+    excuteInitialSearch();
+  }, [dispatch, getRandomYouTubeSearchTopic]);
```
await이 필요한 상황이 아니라서 굳이 async 함수로 래핑할 필요가 없어보입니다. 
```js
useEffect(() => {
  executeSearch(getRandomYoutubeSearchTopic())
}, [dispatch, getRandomYoutubeSearchTopic])
```

pjb6510 commented on src/components/VideoList.jsx:
```diff
@@ -1,17 +1,47 @@
+    if (scrollPosition >= triggerPosition) {
+      loadMoreVideos();
+    }
+  }, 500);
```
기왕이면 이런 상수는 로직 내에 직접 위치시키지 말고 유의미한 이름으로 코드 밖으로 분리시키는게 좋습니다.
여기선 컴포넌트 밖 스코프 정도(대략 7번째줄)에 THROTTLE_TIME 정도의 이름으로 빼면 어떨까 합니다.

pjb6510 commented on src/utils/utils.js:
```diff
@@ -14,6 +14,29 @@ export const debounce = (fn, ms) => {
+  };
+};
+
+export const throttle = (fn, wait) => {
+  let lastCallTime = 0;
+  let timer;
+
+  return (...args) => {
+    const now = Date.now();
+    const remainingTime = wait - (now - lastCallTime);
+
+    if (remainingTime <= 0) {
+      clearTimeout(timer);
+      fn(...args);
+      lastCallTime = Date.now();
+      return;
+    }
+
+    clearTimeout(timer);
+    timer = setTimeout(() => {
+      fn(...args);
+      lastCallTime = Date.now();
+    }, remainingTime);
```
throttle의 옵션으로 { leadingEdge : boolean, trailingEdge : boolean }
을 추가 구현해보세요

SangJin983 committed:
refactor: throttle 함수 리팩토링 및 getRandomIndices 추가
- leading, trailing 옵션을 추가했다

SangJin983 committed:
feat: fetchMockData 함수 추가
- 검색 카테고리별 mockData 추가
- gitIgnore로 mockData가 유출되는 것을 방지

SangJin983 committed:
refactor: excuteSearch, loadMoreVideo 개선
- excuteSearch의 함수명을 보다 일관성 있게 loadVideo로 개선
- useCallback으로 리랜더링시 함수가 재선언되어 생기는 문제를 방지
- THROTTLE_TIME과 같은 식으로, 상수를 컴포넌트 외부로 분리
- setInterval을 활용하여, 주기적으로 스크롤 위치 추적 (최하단에서 스크롤을 움직이지 않을시 무한 스크롤이 작동하지 않는 문제를 해결)
- 일반적이고 다소 모호한 변수명들을 개선

SangJin983 committed:
refactor: config파일을 통한 설정값 관리 및 주석 추가
- config 파일을 따로 만들어 가독성과 유지보수성을 향상시켰다
- 디버깅을 위한 콘솔 로그에 주석을 달아 알아보기 쉽게 만들었다

SangJin983 committed:
refactor: 쓰로틀 기능에서 지연시간에 따른 스크롤 위치추적 문제 해결을 개선
- 기존의 코드는 쓰로틀 타임이 너무 길어(5000ms) 생긴, '사용자가 스크롤을 최하단으로 옮긴 경우 반응하지 않는'문제를 주기적으로 스크롤 위치를 해결하는 setInterval로 해결했으나, 이를 쓰로틀 타임을 줄이는 것(500ms)으로 해결함.

SangJin983 committed:
style: MCOK 오타 수정 (->MOCK)

SangJin983 commented:
### 변경 사항(리뷰 반영 및 추가 개선)
- [리뷰] 코드의 일관성을 위한 네이밍 개선
- [리뷰] 쓰로틀 함수에서 options 구현
- [리뷰] 불피요한 async 래핑 삭제
- [리뷰, 추가] 환경 변수 및 상수의 관리 개선
- [추가] API 할당량 문제 해결을 위한 fetchMockData 함수 구현
- [추가] 디버깅용 콘솔 로그 추가

### 세부 변경 사항
- 네이밍 개선
```VideoSearch.jsx
  // excuteSearch -> loadVideos
  const loadVideos = useCallback(
```
```videoSlice.js
  // results -> searchResult
  initialState: {
    searchResult: [],
```
- 쓰로틀 함수에서 leading, trailing 옵션 구현
```utils.js
export const throttle = (
  // ...,
  options = { leading: true, trailing: false }
) => {
  // ...
    if (remainingTime <= 0) {
      if (options.leading) { 
        // ...
    if (options.trailing) { // ...
```
- api의 환경변수 및 설정값을 관리하기 위한 apiConfig.js 파일 생성
```
### 목적 및 주의 사항
- mockData 사용 여부를 비롯한 다른 환경 변수 및 상수 관리를 용이하게 하기 위해서 해당 파일을 작성했습니다.
- 하드 코딩 된 데이터가 노출 되는 것을 막기 위해 관련 파일은 gitIgonre에 포함 되었습니다.
```
```apiConfig.js
export const API_CONFIG = {
  BASE_URL: import.meta.env.VITE_YOUTUBE_API_BASE_URL,
  API_KEY: import.meta.env.VITE_YOUTUBE_API_KEY,
  USE_MOCK_DATA: import.meta.env.VITE_USE_MOCK_DATA === "true",
};

export const MOCK_DATA_FILE_NAMES = {
  코딩: "coding", // ...
```
- apiConfig 적용
```youtubeApi.js
import { API_CONFIG, MOCK_DATA_FILE_NAMES } from "../config/apiConfig";

const fetchMockData = async (searchQuery) => {
  const fileName = MOCK_DATA_FILE_NAMES[searchQuery];

export const fetchYouTubeSearchResults = async ( // ...
  if (API_CONFIG.USE_MOCK_DATA) {
  // ...

  const params = new URLSearchParams({
    // ...
    key: API_CONFIG.API_KEY,
  });

  const res = await fetch(`${API_CONFIG.BASE_URL}/search?${params}`);
  // ...
```
- fetchMockData 구현에서의 경로설정 및 fetch할 item 선택 로직
```youtubeApi.js
  // 효율적으로 fetch요청할 경로를 구하기 위한 로직
  const fileName = MOCK_DATA_FILE_NAMES[searchQuery];
  if (fileName == null) { // 파일이 없을 경우 분기로 처리
  const res = await fetch(`/mocks/${fileName}.json`);

  // fetch된 비디오 중 필요한 갯수만큼 랜덤으로 반환하기 위한 로직
  const randomIndices = getRandomIndices(0, lastIndex, size);
  const randomItems = randomIndices.map((idx) => data.items[idx]);
```
- getRandomIndices에서의 중복 없는 랜덤 인덱스 선택을 위한 로직
```utils.js
  // 중복된 값을 배제하기 위한 Set 활용
  const result = new Set();

  while (result.size < size) {
    const randomIndex = Math.floor(Math.random() * (max - min + 1)) + min;
    result.add(randomIndex); // 중복된 인덱스의 경우 추가되지 않음
```
- 디버깅을 위한 콘솔 로그들
```
### 사용처와 사용 목적
- fetchMockData, fetchYouTubeSearchResults, loadMoreVideos, handleSubmit에 로그를 출력하는 코드를 추가했습니다.
- 어떤 작업이 호출됐는지, 작업 수행시 입력한 값이나 쓰이는 상태가 어떠한지 확인하기 위함입니다.
- 추후 작업이 전부 끝나면 삭제하도록 하겠습니다. 아래는 실제 쓰인 예시입니다.
```
```VideoSeach.jsx
  // 디버깅을 위한 콘솔 로그(호출한 작업 확인)
  console.log("fetch(mockData)가 실행됐습니다");
```
```VideoList.jsx
      // 디버깅을 위한 콘솔 로그(호출한 작업과 쓰인 상태 및 입력 확인)
      console.log("비디오를 추가합니다. 현재 searchTerm:", searchTerm);
```

### 추후 개선점
- 리뷰를 반영한 실시간 PR 문서 작성(혹은 리플)을 통해, 리뷰어가 보다 용이하게 상황을 파악할 수 있도록 하겠습니다.

### 참고 자료
- [mdn의 Set 설명](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Set)

******

pjb6510 commented on src/components/VideoList.jsx:
```diff
@@ -1,40 +1,54 @@
+      const { items, nextPageToken } = await fetchYouTubeSearchResults(
+        searchTerm,
+        pageToken
+      );
+      const action = appendVideos({ items, nextPageToken });
+      dispatch(action);
+      // 디버깅을 위한 콘솔 로그
+      console.log("비디오를 추가합니다. 현재 searchTerm:", searchTerm);
+    },
+    [fetchYouTubeSearchResults, appendVideos, dispatch]
+  );
 
-  const handleScroll = throttle(() => {
+  const throttledLoadMoreVideos = useCallback(
+    throttle(loadMoreVideos, THROTTLE_TIME),
+    [throttle, loadMoreVideos]
+  );
+
+  const handleScroll = () => {
     const scrollPosition = window.innerHeight + window.scrollY;
     const triggerPosition = document.body.offsetHeight - 500;
 
     // 현재 스크롤 위치가 트리거 위치에 도달했는지 확인
     if (scrollPosition >= triggerPosition) {
-      loadMoreVideos();
+      throttledLoadMoreVideos(searchQuery, pageToken);
     }
-  }, 500);
+  };
 
-  // handleScroll 이벤트 핸들러를 스크롤 이벤트에 등록
   useEffect(() => {
+    // handleScroll 이벤트 핸들러를 스크롤 이벤트에 등록
     window.addEventListener("scroll", handleScroll);
     // 컴포넌트 언마운트 시 이벤트 핸들러 해제
-    return () => window.removeEventListener("scroll", handleScroll);
+    return () => {
+      window.removeEventListener("scroll", handleScroll);
+    };
   }, [handleScroll]);
```
여기서 useCallback을 쓴 의도를 모르겠습니다.
useCallback이 아무런 역할도 못하는 것 같습니다.

useCallback에 의해 래핑되어있는 `loadMoreVideos`와 
`throttledLoadMoreVideos`는 memoized 되겠지만 
정작, 그걸 호출하는 handleScroll은 useCallback 래핑되어 있지 않기 때문에 컴포넌트가 리렌더 될 때마다 새로 선언될 것으로 보입니다.

handleScroll은 useEffect에 dependencies로 들어가 있기 때문에 useEffect는 매 렌더링마다 재실행되는 셈이 되겠네요.

SangJin983 replied:
이 부분은 본래 throttle 함수가 초기화(타이머, 콜타임) 되는 문제를 어떻게 해결하는게 좋을지, 논의를 가져보고자 했었습니다.

문제 설명:
- 리랜더 시 함수들이 재선언 되면서, throttle 함수가 본래 관리해야 할, 타이머와 콜타임이 초기화 되어서, 원래 생각했던 쓰로틀링이 제대로 동작하지 않음을 확인

해결책(임시):
- throttle 함수의 타이머와 콜타임이 초기화되지 않게 하기 위해서, `throttleLoadMoreVideos`와 `loadMoreVideos`를 useCallback으로 감싸줌

해당 부분을 이렇게 처리하는게, useCallback을 사용하는 목적에 맞는지 의문이 들었기 때문에, 추가적인 작업을 보류해 놨습니다. (아래에 말씀하신대로 `loadVideos`의 useCallback은 목적에 맞다고 생각합니다)
연구하다가 뾰족한 수를 못 찾아서 다른 과제부터 해결하다가... 잊어버렸습니다😭.

아래에 말씀하신 자료도 같이 연구하고, 적절하게 수정해보겠습니다.

별개로 PR에서 '스스로의 작업에 미진한 부분이나 의문거리'를 적어놓는게, 적절한 PR 양식에 맞는지 약간 의문이 있는데 (PR에다가 적는게 올바른지, 아니면 깃헙의 뭔가 다른 기능을 써서 하는지) 문제가 없다면, 추후에는 그런 부분도 PR에 반영하도록 하겠습니다.

요약:
1. 해당 부분은 리랜더링시 throttle 함수의 타이머, 콜타임이 초기화되는 문제(쓰로틀링이 정상 동작하지 않음)를 해결하기 위함
2. 그러나 작성자는 이런 방식이 useCallback의 본래 사용목적에 맞는지 의문이 들어서, 작업을 보류함
3. 좀더 연구하면서 적절한 답을 찾아보겠음

pjb6510 commented on src/components/VideoSearch.jsx:
```diff
@@ -8,35 +8,39 @@ const VideoSearch = () => {
+      const action = setVideos({
+        items,
+        nextPageToken,
+        searchTerm,
+      });
+      dispatch(action);
+    },
+    [fetchYouTubeSearchResults, setVideos, dispatch]
+  );
 
   const handleSubmit = async (e) => {
     e.preventDefault(); // 새로고침 방지
     if (searchTerm.trim()) {
-      excuteSearch(searchTerm);
+      loadVideos(searchTerm);
+      // 디버깅을 위한 콘솔 로그
+      console.log("검색한 비디오를 불러옵니다. searchTerm:", searchTerm);
+
       setSearchTerm("");
     }
   };
 
   // 최초 랜덤검색 기능 구현
   useEffect(() => {
-    const excuteInitialSearch = async () => {
-      const randomTopic = getRandomYouTubeSearchTopic();
-      excuteSearch(randomTopic);
-    };
-
-    excuteInitialSearch();
-  }, [dispatch, getRandomYouTubeSearchTopic]);
+    const randomTopic = getRandomYouTubeSearchTopic();
+    // 디버깅을 위한 콘솔 로그
+    console.log("randomTopic:", randomTopic);
+    loadVideos(randomTopic);
+  }, [loadVideos, getRandomYouTubeSearchTopic]);
```
useEffect의 dependencies로 들어가는 함수의 재선언을 제한하기 위해서라면 이런 식으로 쓰이는게 맞습니다.
- getRandomYouTubeSearchTopic 은 재선언되지 않으니 논외입니다.
- loadVideos는 fetchYouTubeSearchResults, setVideos, dispatch에 의해 재선언 됩니다.
  - (사실 세 함수 모두 재선언될 여지는 없어보입니다. useEffect는 렌더링 최초에만 호출 되겠네요.)

따라서
1. fetchYouTubeSearchResults, setVideos, dispatch의 재선언
2. 그에 따른 loadVideos의 재선언
3. 그에 따른 useEffect 콜백 호출이 될 것 같습니다.

이러한 케이스에서 위와 같은 연쇄반응을 이해하는 것이 중요합니다.
위의 handleScroll도 참고하여 수정해보세요.

pjb6510 replied:
https://rinae.dev/posts/a-complete-guide-to-useeffect-ko/

리덕스를 만들고 메타에서 리액트에 기여했던 dan abramov라는 개발자의 useEffect에 관한 글입니다.
한번 쭉 읽어보세요.

SangJin983 replied:
읽고 연구해보겠습니다.

pjb6510 commented:
전반적으로 잘 하셨습니다. 마무리하고 다음 과제로 넘어가시죠 👍
